### ISSUE TITLE
kube-dns never in ready state with etcd-tls (always 2/3, 1/3)

### ISSUE THREAD

**Issue by gree-gorey** [projectcalico/canal#81]
When create canal with `canal_etcd_tls.yaml` kube-dns pod never reaches ready state, only 2/3 or 1/3.
When deploying canal without tls, everything works fine.
We have 2-node etcd cluster running outside of kube.

kubelet version: 1.6.6
docker version: 1.12

`kubectl logs kube-dns-692378583-xdxkh kubedns -n kube-system` output:

```
I0706 09:52:22.211047       1 dns.go:48] version: 1.14.1-16-gff416ee
I0706 09:52:22.212257       1 server.go:70] Using configuration read from directory: /kube-dns-config with period 10s
I0706 09:52:22.212432       1 server.go:113] FLAG: --alsologtostderr="false"
I0706 09:52:22.212452       1 server.go:113] FLAG: --config-dir="/kube-dns-config"
I0706 09:52:22.212472       1 server.go:113] FLAG: --config-map=""
I0706 09:52:22.212505       1 server.go:113] FLAG: --config-map-namespace="kube-system"
I0706 09:52:22.212523       1 server.go:113] FLAG: --config-period="10s"
I0706 09:52:22.212535       1 server.go:113] FLAG: --dns-bind-address="0.0.0.0"
I0706 09:52:22.212547       1 server.go:113] FLAG: --dns-port="10053"
I0706 09:52:22.212565       1 server.go:113] FLAG: --domain="cluster.local."
I0706 09:52:22.212623       1 server.go:113] FLAG: --federations=""
I0706 09:52:22.212644       1 server.go:113] FLAG: --healthz-port="8081"
I0706 09:52:22.212654       1 server.go:113] FLAG: --initial-sync-timeout="1m0s"
I0706 09:52:22.212659       1 server.go:113] FLAG: --kube-master-url=""
I0706 09:52:22.212671       1 server.go:113] FLAG: --kubecfg-file=""
I0706 09:52:22.212721       1 server.go:113] FLAG: --log-backtrace-at=":0"
I0706 09:52:22.212730       1 server.go:113] FLAG: --log-dir=""
I0706 09:52:22.212743       1 server.go:113] FLAG: --log-flush-frequency="5s"
I0706 09:52:22.212754       1 server.go:113] FLAG: --logtostderr="true"
I0706 09:52:22.212765       1 server.go:113] FLAG: --nameservers=""
I0706 09:52:22.212802       1 server.go:113] FLAG: --stderrthreshold="2"
I0706 09:52:22.212820       1 server.go:113] FLAG: --v="2"
I0706 09:52:22.212831       1 server.go:113] FLAG: --version="false"
I0706 09:52:22.212844       1 server.go:113] FLAG: --vmodule=""
I0706 09:52:22.212966       1 server.go:176] Starting SkyDNS server (0.0.0.0:10053)
I0706 09:52:22.213179       1 server.go:198] Skydns metrics enabled (/metrics:10055)
I0706 09:52:22.213204       1 dns.go:147] Starting endpointsController
I0706 09:52:22.213211       1 dns.go:150] Starting serviceController
I0706 09:52:22.213298       1 logs.go:41] skydns: ready for queries on cluster.local. for tcp://0.0.0.0:10053 [rcache 0]
I0706 09:52:22.213326       1 logs.go:41] skydns: ready for queries on cluster.local. for udp://0.0.0.0:10053 [rcache 0]
I0706 09:52:22.713375       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
...
E0706 09:52:52.214678       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Service: Get https://10.96.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout
```

`kubectl logs kube-dns-692378583-xdxkh dnsmasq -n kube-system` output:

```
I0706 09:55:56.298941       1 main.go:76] opts: {{/usr/sbin/dnsmasq [-k --cache-size=1000 --log-facility=- --server=/cluster.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053] true} /etc/k8s/dns/dnsmasq-nanny 10000000000}
I0706 09:55:56.299052       1 nanny.go:86] Starting dnsmasq [-k --cache-size=1000 --log-facility=- --server=/cluster.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053]
I0706 09:55:56.372491       1 nanny.go:111] 
W0706 09:55:56.372511       1 nanny.go:112] Got EOF from stdout
I0706 09:55:56.372525       1 nanny.go:108] dnsmasq[13]: started, version 2.76 cachesize 1000
I0706 09:55:56.372534       1 nanny.go:108] dnsmasq[13]: compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-DNSSEC loop-detect inotify
I0706 09:55:56.372543       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0706 09:55:56.372550       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0706 09:55:56.372552       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain cluster.local 
I0706 09:55:56.372572       1 nanny.go:108] dnsmasq[13]: reading /etc/resolv.conf
I0706 09:55:56.372585       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0706 09:55:56.372589       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0706 09:55:56.372591       1 nanny.go:108] dnsmasq[13]: using nameserver 127.0.0.1#10053 for domain cluster.local 
I0706 09:55:56.372593       1 nanny.go:108] dnsmasq[13]: using nameserver 8.8.8.8#53
I0706 09:55:56.372596       1 nanny.go:108] dnsmasq[13]: using nameserver 192.168.0.1#53
I0706 09:55:56.372599       1 nanny.go:108] dnsmasq[13]: read /etc/hosts - 7 addresses
```

`kubectl logs kube-dns-692378583-xdxkh sidecar -n kube-system` output:

```
ERROR: logging before flag.Parse: I0706 09:44:15.355606       1 main.go:48] Version v1.14.1-16-gff416ee
ERROR: logging before flag.Parse: I0706 09:44:15.355794       1 server.go:45] Starting server (options {DnsMasqPort:53 DnsMasqAddr:127.0.0.1 DnsMasqPollIntervalMs:5000 Probes:[{Label:kubedns Server:127.0.0.1:10053 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:1} {Label:dnsmasq Server:127.0.0.1:53 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:1}] PrometheusAddr:0.0.0.0 PrometheusPort:10054 PrometheusPath:/metrics PrometheusNamespace:kubedns})
ERROR: logging before flag.Parse: I0706 09:44:15.356011       1 dnsprobe.go:75] Starting dnsProbe {Label:kubedns Server:127.0.0.1:10053 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:1}
ERROR: logging before flag.Parse: I0706 09:44:15.356147       1 dnsprobe.go:75] Starting dnsProbe {Label:dnsmasq Server:127.0.0.1:53 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:1}
ERROR: logging before flag.Parse: W0706 09:48:05.395021       1 server.go:64] Error getting metrics from dnsmasq: read udp 127.0.0.1:55328->127.0.0.1:53: read: connection refused
```

**Comment by mastanggt** [comment]
+1
We have the same problem.

**Comment by caseydavenport** [comment]
Hm, so this line indicates that the DNS pod is failing to reach the apiserver:

```
E0706 09:52:52.214678       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Service: Get https://10.96.0.1:443/api/v1/services?resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout
```

This could be due to a number of things, but most likely culprits:
- kube-proxy is not running / healthy
- flannel is not running / healthy / configured properly
- the Calico CNI plugin is not configured properly, or Calico policy is preventing the traffic.

Can you check if:
- pod->pod traffic works (not through a Service)
- pod->master traffic works (not through a Service)

That will help decide if it's a kube-proxy issue or a canal issue.

**Comment by gree-gorey** [comment]
The problem was with misconfiguration of etcd tls certificates.

But when we configured them properly, canal-configure job starts failing -- it could not connect to etcd cluster.
Then we change this block from [canal_etcd_tls.yaml](https://github.com/projectcalico/canal/blob/master/k8s-install/canal_etcd_tls.yaml#L313)
```yaml
command:
  - "etcdctl"
  - "--ca-file=/calico-secrets/etcd-ca"
  - "set"
  - "/coreos.com/network/config"
  - '{ "Network": "172.30.0.0/16", "Backend": {"Type": "vxlan"} }'
```

to this:
```yaml
command:
  - "etcdctl"
  - "--cert-file=/calico-secrets/etcd-cert"
  - "--key-file=/calico-secrets/etcd-key"
  - "--ca-file=/calico-secrets/etcd-ca"
  - "--no-sync"
  - "set"
  - "/coreos.com/network/config"
  - '{ "Network": "172.30.0.0/16", "Backend": {"Type": "vxlan"} }'
```

and after that everything starts working.
btw, was it the right thing to do?

For those interested, [our file for canal etcd tls](https://gist.github.com/gree-gorey/00fde48e79da04dd11997946423e5ed6)

**Comment by caseydavenport** [comment]
@gree-gorey those changes look right to me - glad that got it working!

Would you be open to submitting a PR to upstream those changes?

**Comment by gree-gorey** [comment]
@caseydavenport yeah, sure. Submitted PR.

**Comment by caseydavenport** [comment]
Thanks! Fixed in #82
